File: src/providers/openai-cua/rules/structured-step-metadata-not-text-json.ts
/**
 * Flags a function that both (a) searches free text for a brace via
 * indexOf/lastIndexOf, and (b) calls JSON.parse on a slice/substring of
 * that text — manual brace-hunting JSON extraction instead of the
 * Responses API's structured tool/function output.
 */
const BRACE_SEARCH_METHODS = new Set(['indexOf', 'lastIndexOf']);
const SLICE_METHODS = new Set(['slice', 'substring', 'substr']);

const rule = {
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Step metadata must come from structured tool output, not brace-hunting in free text',
      category: 'correctness',
      rationale:
        'Manually locating a JSON object inside a free-text model message (via indexOf/lastIndexOf plus a slice) is fragile by construction: it breaks if the model adds trailing commentary, wraps the JSON in a markdown fence, or reorders fields. The Responses API supports function tools and structured output specifically so required metadata is schema-validated by the API instead of regex/brace-scraped out of arbitrary text.',
      docsUrl: 'https://developers.openai.com/api/docs/guides/tools-computer-use',
      recommended: true,
    },
    messages: {
      textJsonExtraction:
        'This function locates a JSON object in free text via indexOf/lastIndexOf and parses a slice of it — use a function tool or structured output instead of brace-hunting in the model\'s text message.',
    },
  },
  create(context: any) {
    type FnState = { sawBraceSearch: boolean; sawJsonParseOnSlice: boolean; reportNode: any };
    const stack: any[] = [];
    const states = new Map<any, FnState>();

    function ensureState(fn: any): FnState {
      let s = states.get(fn);
      if (!s) {
        s = { sawBraceSearch: false, sawJsonParseOnSlice: false, reportNode: null };
        states.set(fn, s);
      }
      return s;
    }

    function top(): any {
      return stack[stack.length - 1];
    }

    function pushScope(node: any) {
      stack.push(node);
      ensureState(node);
    }

    function popScope() {
      stack.pop();
    }

    function isBraceSearchCall(node: any): boolean {
      if (node?.type !== 'CallExpression') return false;
      const callee = node.callee;
      if (callee?.type !== 'MemberExpression') return false;
      if (!BRACE_SEARCH_METHODS.has(callee.property?.name)) return false;
      const arg = node.arguments?.[0];
      return arg?.type === 'Literal' && typeof arg.value === 'string' && arg.value.includes('{');
    }

    function isJsonParseOnSliceCall(node: any): boolean {
      if (node?.type !== 'CallExpression') return false;
      const callee = node.callee;
      if (callee?.type !== 'MemberExpression') return false;
      if (callee.object?.type !== 'Identifier' || callee.object.name !== 'JSON') return false;
      if (callee.property?.name !== 'parse') return false;
      const arg = node.arguments?.[0];
      if (arg?.type !== 'CallExpression') return false;
      const argCallee = arg.callee;
      return argCallee?.type === 'MemberExpression' && SLICE_METHODS.has(argCallee.property?.name);
    }

    return {
      Program(node: any) {
        pushScope(node);
      },
      'Program:exit'() {
        for (const state of states.values()) {
          if (state.sawBraceSearch && state.sawJsonParseOnSlice) {
            context.report({ node: state.reportNode, messageId: 'textJsonExtraction' });
          }
        }
      },

      FunctionDeclaration(node: any) {
        pushScope(node);
      },
      'FunctionDeclaration:exit'() {
        popScope();
      },
      FunctionExpression(node: any) {
        pushScope(node);
      },
      'FunctionExpression:exit'() {
        popScope();
      },
      ArrowFunctionExpression(node: any) {
        pushScope(node);
      },
      'ArrowFunctionExpression:exit'() {
        popScope();
      },

      CallExpression(node: any) {
        const fn = top();
        if (!fn) return;
        const state = ensureState(fn);

        if (isBraceSearchCall(node)) {
          state.sawBraceSearch = true;
          if (!state.reportNode) state.reportNode = node;
        }
        if (isJsonParseOnSliceCall(node)) {
          state.sawJsonParseOnSlice = true;
          if (!state.reportNode) state.reportNode = node;
        }
      },
    };
  },
};

export const openaiCuaStructuredStepMetadataNotTextJsonRule = rule;

File: src/providers/openai-cua/rules/python/structured_step_metadata_not_text_json.py
"""openai-cua-structured-step-metadata-not-text-json

Parity with the JS rule: flags a function that both (a) searches free text for
a brace via str.find/rfind/index/rindex, and (b) calls json.loads on a slice
of that text — manual brace-hunting JSON extraction instead of a function
tool / structured output.
"""
from __future__ import annotations

import ast
import sys
from pathlib import Path

from ast_utils import enclosing_scope_map, loc

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))

RULE_KEY = "openai-cua-structured-step-metadata-not-text-json"

BRACE_SEARCH_METHODS = frozenset({"find", "rfind", "index", "rindex"})


def _is_brace_search_call(node: ast.AST) -> bool:
    if not isinstance(node, ast.Call) or not isinstance(node.func, ast.Attribute):
        return False
    if node.func.attr not in BRACE_SEARCH_METHODS:
        return False
    if not node.args:
        return False
    arg = node.args[0]
    return isinstance(arg, ast.Constant) and isinstance(arg.value, str) and "{" in arg.value


def _is_json_loads_on_slice(node: ast.AST) -> bool:
    if not isinstance(node, ast.Call):
        return False
    func = node.func
    is_loads = (isinstance(func, ast.Attribute) and func.attr == "loads") or (
        isinstance(func, ast.Name) and func.id == "loads"
    )
    if not is_loads or not node.args:
        return False
    arg = node.args[0]
    return isinstance(arg, ast.Subscript) and isinstance(arg.slice, ast.Slice)


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    scope_of = enclosing_scope_map(tree)
    brace_search_scopes: set[int] = set()
    json_slice_scopes: set[int] = set()
    report_node_by_scope: dict[int, ast.AST] = {}

    for node in ast.walk(tree):
        scope = scope_of.get(id(node), tree)
        scope_id = id(scope)

        if _is_brace_search_call(node):
            brace_search_scopes.add(scope_id)
            report_node_by_scope.setdefault(scope_id, node)
        if _is_json_loads_on_slice(node):
            json_slice_scopes.add(scope_id)
            report_node_by_scope.setdefault(scope_id, node)

    out: list[dict] = []
    for scope_id in brace_search_scopes & json_slice_scopes:
        node = report_node_by_scope[scope_id]
        line, col, _, _ = loc(node)
        out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})
    return out
